File: app/src/main/java/threadpoolexec/com/threadpoolexecutor/DI/NetworksModule.java
package threadpoolexec.com.threadpoolexecutor.DI;

/**
 * Dagger Provider component
 * (a singleton object which is needed)
 */

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.util.concurrent.Executors;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
public class NetworksModule {
    private String urlPath;

    public NetworksModule(String urlPath) {
        this.urlPath = urlPath;
    }

    @Singleton
    @Provides
    public Gson provideGson() {
        GsonBuilder builder = new GsonBuilder();
        return builder.create();
    }

    @Singleton
    @Provides
    public Retrofit provideRetrofit(Gson gson) {
        Retrofit retrofit = new Retrofit.Builder()
                //this urlpath is unused(Retrofit allows dynamic urls through @Url)
                .baseUrl(urlPath)
                //retrofit callback hits on main thread...use pool
                .callbackExecutor(Executors.newSingleThreadExecutor())
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
        return retrofit;
    }
}

File: app/src/main/java/threadpoolexec/com/threadpoolexecutor/DI/NetworksModule.kt
package threadpoolexec.com.threadpoolexecutor.DI

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import dagger.Module
import dagger.Provides
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.Executors


@Module
class NetworksModule(private val urlPath: String) {

    @Provides
    fun provideGson(): Gson = GsonBuilder().create()

    @Provides
    fun provideRetrofit(gson: Gson): Retrofit =
            Retrofit.Builder()
                    .baseUrl(urlPath)
                    .callbackExecutor(Executors.newSingleThreadExecutor())
                    .addConverterFactory(GsonConverterFactory.create(gson))
                    .build()
}